pkg/maglev/maglev.go

```diff
@@ -7,77 +7,135 @@ import (
+		if cfg.MaglevTableSize == prime {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil, fmt.Errorf("Invalid value for --%s: %d, supported values are: %v",
+			MaglevTableSizeName, cfg.MaglevTableSize, supportedPrimes)
+	}
+
+	seed := cfg.MaglevHashSeed
 	d, err := base64.StdEncoding.DecodeString(seed)
 	if err != nil {
-		return fmt.Errorf("Cannot decode base64 Maglev hash seed %q: %w", seed, err)
+		return nil, fmt.Errorf("cannot decode base64 Maglev hash seed %q: %w", seed, err)
 	}
 	if len(d) != 12 {
-		return fmt.Errorf("Decoded hash seed is %d bytes (not 12 bytes)", len(d))
+		return nil, fmt.Errorf("decoded hash seed is %d bytes (not 12 bytes)", len(d))
 	}
 
-	seedMurmur = uint32(d[0])<<24 | uint32(d[1])<<16 | uint32(d[2])<<8 | uint32(d[3])
-
-	SeedJhash0 = uint32(d[4])<<24 | uint32(d[5])<<16 | uint32(d[6])<<8 | uint32(d[7])
-	SeedJhash1 = uint32(d[8])<<24 | uint32(d[9])<<16 | uint32(d[10])<<8 | uint32(d[11])
+	seedMurmur := uint32(d[0])<<24 | uint32(d[1])<<16 | uint32(d[2])<<8 | uint32(d[3])
+	SeedJhash0 := uint32(d[4])<<24 | uint32(d[5])<<16 | uint32(d[6])<<8 | uint32(d[7])
+	SeedJhash1 := uint32(d[8])<<24 | uint32(d[9])<<16 | uint32(d[10])<<8 | uint32(d[11])
 
-	// Allocate this ahead of time to avoid expensive allocations inside
-	// getPermutation().
-	permutation = make([]uint64, derivePermutationSliceLen(m))
+	minSize := derivePermutationSliceLen(uint64(cfg.MaglevTableSize))
 
-	return nil
+	return &Maglev{
+		Config:       cfg,
+		seedMurmur:   seedMurmur,
+		SeedJhash0:   SeedJhash0,
+		SeedJhash1:   SeedJhash1,
+		backendInfos: make([]backendInfo, 0, 1024),
+		permutations: make([]uint64, minSize),
+		wp:           workerpool.New(runtime.NumCPU()),
+	}, nil
 }
 
-func getOffsetAndSkip(backend string, m uint64) (uint64, uint64) {
-	h1, h2 := murmur3.Hash128([]byte(backend), seedMurmur)
-	offset := h1 % m
-	skip := (h2 % (m - 1)) + 1
-
-	return offset, skip
-}
+func (ml *Maglev) getPermutation(backends []backendInfo, m int, numCPU int) []uint64 {
+	if len(backends) == 0 {
+		return nil
+	}
 
-func getPermutation(backends []string, m uint64, numCPU int) []uint64 {
 	// The idea is to split the calculation into batches so that they can be
 	// concurrently executed. We limit the number of concurrent goroutines to
 	// the number of available CPU cores. This is because the calculation does
 	// not block and is completely CPU-bound. Therefore, adding more goroutines
 	// would result into an overhead (allocation of stackframes, stress on
 	// scheduling, etc) instead of a performance gain.
-
 	bCount := len(backends)
-	if size := uint64(bCount) * m; size > uint64(len(permutation)) {
+
+	// As the permutations array is large and often used, we'll keep a single
+	// instance around and reuse it.
+	if size := bCount * m; size > len(ml.permutations) {
 		// Reallocate slice so we don't have to allocate again on the next
 		// call.
-		permutation = make([]uint64, size)
+		ml.permutations = make([]uint64, size)
```

Comment:
If `m` cannot be increased at runtime, the only situation when the reallocation beyond the initial size is needed is when `bCount` exceeds `m/100`. Could we add a warning log here to indicate it?

The reason why I'm so concerned about this are the exemplar values in the comments above `derivePermutationsSliceLen` saying that the permutation can grow up to 1GB+ (for non-default `m`). I'm not sure if the computation really remains feasible at such scales, but even a single reallocation would cause a memory usage spike by 1GB (until the previous slice is garbage collected), which may be problematic with tight memory limits applied by k8s users to control costs.

Btw. since `m == uint64(ml.Config.MaglevTableSize)`, perhaps we can remove it from the signature of `getPermutation`?